src/graph/LdapAuthenticator.cpp

```diff
@@ -0,0 +1,290 @@
+            !FLAGS_ldap_bindpasswd.empty() ||
+            !FLAGS_ldap_searchattribute.empty() ||
+            !FLAGS_ldap_searchfilter.empty()) {
+            return Status::Error("LDAP authentication direct bind mode and search bind mode"
+                                 " parameters cannot be mixed.");
+        }
+    // Here search bind mode must be set FLAGS_ldap_basedn
+    } else if (FLAGS_ldap_basedn.empty()) {
+        return Status::Error("LDAP authentication requires argument \"ldap_prefix\", "
+                             "\"ldap_suffix\", or \"ldap_basedn\" to be set.");
+    }
+
+    /**
+     * Search bind mode can either use FLAGS_ldap_searchattribute or
+     * FLAGS_ldap_searchfilter, but can't use both. FLAGS_Ldap_searchattribute
+     * default is "uid". FLAGS_Ldap_searchfilter is more flexible search filtes
+     * than FLAGS_Ldap_searchattribute.
+     */
+    if (!FLAGS_ldap_searchattribute.empty() && !FLAGS_ldap_searchfilter.empty()) {
+        return Status::Error("LDAP authentication cannot use ldap_searchattribute "
+                             "together with ldap_searchfilter.");
+    }
+
+    return Status::OK();
+}
+
+
+Status initLDAPConnection() {
+    int ldap_version = LDAP_VERSION3;
+
+    int ret  = ldap_create(&ldap_);
+    if (ret != LDAP_SUCCESS) {
+        return Status::Error("Init LDAP failed.");
+    }
+    ldap_->ld_options.ldo_defport = FLAGS_ldap_port;
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_HOST_NAME, FLAGS_ldap_server);
+    if (ret != LDAP_SUCCESS) {
+        ldap_ld_free(ldsp_, 1, NULL, NULL);
+        return Status::Error("Init LDAP failed.");
+    }
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_PROTOCOL_VERSION, &ldap_version);
+    if (ret != LDAP_SUCCESS) {
+        ldap_unbind(ldap_);
+        return Status::Error("Set LDAP protocol version failed");
+    }
+
+    if (FLAGS_ldap_tls) {
+        ret = ldap_start_tls_s(ldap_, NULL, NULL));
+        if (ret != LDAP_SUCCESS) {
+            ldap_unbind(ldap_);
+            return Status::Error("Start LDAP TLS session failed");
+        }
+    }
+
+    return Status::OK();
+}
+
+
+std::string LdapAuthenticator::buildFilter() {
+    std::string filter;
+    if (!FLAGS_ldap_searchfilter.empty()) {
+        filter = buildSearchFilter();
+    } else if (!FLAGS_ldap_searchattribute.empty()) {
+        filter = folly::stringPrintf("(%s=%s)", FLAGS_ldap_searchattribute.c_str(),
+                                      userName_.c_str());
+    } else {
+        filter = folly::stringPrintf("(uid=%s)", userName_.c_str());
+    }
+    return filter;
+}
+
+
+std::string LdapAuthenticator::buildSearchFilter() {
+    std::string result  = FLAGS_ldap_searchfilter;
+    std::string ph = "$username";
+    auto len = ph.length();
+
+    std::string::size_type  pos = 0;
+    while ((pos = result.find(ph)) != std::string::npos) {
+        result.replace(pos, len, userName_);
+    }
+
+    return result;
+}
+
+
+Status<bool> LdapAuthenticator::searchBindAuth() {
```

Comment:
If I were right, there is no such thing by `Status<bool>`.

Reply:
My mistake, it  should be StatusOr<bool>.